source-code/core/src/lint/createLintRule.ts

```diff
@@ -0,0 +1,56 @@
+import type { LintRule } from "./rule.js"
+
+/**
+ * A utility function to create a lint rule.
+ *
+ * @example
+ * const myRule = createLintRule({ id: "example.rule" }, (args) => {
+ *   // Implement your rule logic here
+ * });
+ *
+ * // using the rule
+ * myRule("error");
+ *
+ * // you can also pass settings to the rule
+ * const myRule2 = createLintRule<{ strict: boolean }>({ id: "example.rule" }, (args) => {
+ *   // Implement your rule logic here
+ * });
+ *
+ * // settings must be provided when using the rule
+ * myRule2("error", { strict: true });
+ */
+export const createLintRule = <Settings extends object | undefined = undefined>(
+	args: { id: LintRule["id"] },
+	setup: (
```

Comment:
should `setup` be integrated into `args`?

Reply:
Yes. I'll change it!

----

While writing down the arguments I came to believe that you are right.

`args` = static "config". i didn't want to name it config though because setup takes the inlang config as argument.
`setup` = dynamic things that actually make up a rule. 

```ts
export type LintRule = {
	id: `${string}.${string}` // <-- static
	level: "error" | "warn"   // <-- static
	setup: Function           // <-- dynamic
}
```

### Problems with the current solution

1. Supossing that a lint rule will be extended with more dynamic "arguments" in the future, using `setup` directly will require us to introduce a third positional argument (we are not using the object argument pattern). 

2. `args` is not clear.

### Proposal 

YEAH NO. Abort the mission. 

```ts
export const createLintRule = <Settings extends Record<string, unknown> | undefined = undefined>(
	static: { id: LintRule["id"] },
	args: {
           setup: (
		args: Parameters<LintRule["setup"]>[0] & {
			settings: Settings
		},
	)} => ReturnType<LintRule["setup"]>,
): ConfigureLintRuleFunction<Settings>
```

Reply:
Wuuups I know why I did this. We should probably go with `static` and the second argument being `callback`. 

If we put `setup` into `args`, a createLintRule can't have shared state anymore. We would end up with this issue again https://github.com/inlang/inlang/issues/457#issue-1631994310. 

```ts
const myRule = createLintRule({
id: "example.rule",
setup: () => {
	return {
		visitors: {
			Resource: () => {},
		},
	}
},
})
```

The current design is read as "the left argument is static config, the right one is a callback for dynamic properties"

```ts
const myRule = createLintRule({ id: "example.rule" }, async () => {

// shared global state
const x = "hello world"

// or initialize an API .
const API = await Grammarly()

return {
	visitors: {
		Resource: () => {},
	},
}
})
```

I tried to remove the `setup` entirely from `LintRule` but a lint rule must have an async callback property that the linter executes. Otherwise, execution must happen in `defineConfig()` which we want to avoid. 

Given that all dynamic "properties" of a lint rule must be returned by setup, I reverse my previous comment. Args and setup should be de-coupled. All dynamic properties will be behind `setup`, thus we are not running into the issue of requiring additional arguments for the `createLintRule` function in the future. 

BUT `args` is not optimal. `static` or `config` would be suitable. Unfortunately, `static` is a reserved keyword and `config` might be confused with the inlang config. 

@ivanhofer ideas for a better name?

Reply:
I just realized, while laying in bed, that ofc we can have shared state if everything dynamic happens in the setup function. 

```diff
const myRule = createLintRule({
id: "example.rule",
setup: async () => {
+       const x = "here is shared state"
	return {
		visitors: {
			Resource: () => {},
		},
	}
},
})
```

For the sake of keeping it simple and future-proof, I now (again :D) vote for merging `args` and `config`.

Reply:
I would merge them!

Reply:
Done